
<h1 align="center"> Segmenting and Clustering Neighbourhoods - Toronto </h1>

I have put all the 3 question in this notebook. Every question has a header before it.

In [1]:
import pandas as pd

# Question 1

## Importing Data

In [9]:
!pip install bs4

     |████████████████████████████████| 122kB 18.7MB/s eta 0:00:01
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/a0/b0/b2/4f80b9456b87abedbc0bf2d52235414c3467d8889be38dd472
Successfully built bs4


In [26]:
import requests
from bs4 import BeautifulSoup

In [49]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
soup = BeautifulSoup(wiki_url.text, 'html5lib')

In [50]:
table_contents=[]
table=soup.find('table')

for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['Postal Code'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)


In [51]:
# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})


In [52]:
df[df['Borough'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

In [53]:
df[df['Neighborhood'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

Grouping the records based on Postal Code

In [54]:
df = df.groupby(['Postal Code']).first()

Checking for number of records where Neighbourhood is "Not assigned"

In [55]:
df.Neighborhood.str.count("Not assigned").sum()

0

In [56]:
df = df.reset_index()
df

Postal Code              Borough  \
0           M1B          Scarborough   
1           M1C          Scarborough   
2           M1E          Scarborough   
3           M1G          Scarborough   
4           M1H          Scarborough   
..          ...                  ...   
98          M9N                 York   
99          M9P            Etobicoke   
100         M9R            Etobicoke   
101         M9V            Etobicoke   
102         M9W  Etobicoke Northwest   

                                          Neighborhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Port Union, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
..                                                 ...  
98                                              Weston  
99                                           Westmount  
100  Kingsview Village, St. Phillips, Martin Grove ...  
101  South Steeles, Silverstone, Humbergate, Jamest...  
102  Clairville, Humberwood, Woodbine Downs, West H...  

[103 rows x 3 columns]

In [57]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

# Question 2

Installing geocoder

In [42]:
!pip install geocoder

     |████████████████████████████████| 102kB 20.3MB/s ta 0:00:01


In [43]:
import geocoder # import geocoder

geocoder wasn't working

Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [44]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [45]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [58]:
df.dtypes

Postal Code     object
Borough         object
Neighborhood    object
dtype: object

In [59]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [60]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code              Borough  \
0           M1B          Scarborough   
1           M1C          Scarborough   
2           M1E          Scarborough   
3           M1G          Scarborough   
4           M1H          Scarborough   
..          ...                  ...   
98          M9N                 York   
99          M9P            Etobicoke   
100         M9R            Etobicoke   
101         M9V            Etobicoke   
102         M9W  Etobicoke Northwest   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724  
101  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
102  Clairville, Humberwood, Woodbine Downs, West H...  43.706748 -79.594054  

[103 rows x 5 columns]

In [61]:
combined_data.shape

(103, 5)

**Solution:** We get 103 rows as expected when we do a inner join, so we have good data.

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [65]:
!pip install geopy

     |████████████████████████████████| 112kB 22.6MB/s eta 0:00:01


In [66]:
from geopy.geocoders import Nominatim 

In [67]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [68]:
import folium

In [70]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighborhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [77]:
CLIENT_ID = 'IEO3BCOMPMVDENSQXFGQOO21FUGTKIV1ZX01F4LNKLV0PK25' # your Foursquare ID
CLIENT_SECRET = 'XNJPG4KKQYJR2ZJM1PFS1FBAOM3APZ32TT2KDDQ3QNBPWCV0' # your Foursquare Secret
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: IEO3BCOMPMVDENSQXFGQOO21FUGTKIV1ZX01F4LNKLV0PK25
CLIENT_SECRET:XNJPG4KKQYJR2ZJM1PFS1FBAOM3APZ32TT2KDDQ3QNBPWCV0


Next, we create a function to get all the venue categories in Toronto

In [88]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [89]:
venues_in_toronto = getNearbyVenues(combined_data['Neighborhood'], combined_data['Latitude'], combined_data['Longitude'])

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Don Mills South
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
The Danforth  East
The Danforth West, Riverdale


In [90]:
venues_in_toronto.shape

(1273, 5)

So we have 1273 records and 5 columns. Checking sample data


In [91]:
venues_in_toronto.head()

Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                                              Venue  \
0              -79.194353                                            Wendy’s   
1              -79.160497                              Royal Canadian Legion   
2              -79.160497  SEBS Engineering Inc. (Sustainable Energy and ...   
3              -79.188711                                     RBC Royal Bank   
4              -79.188711                                  G & G Electronics   

               Venue Category  
0        Fast Food Restaurant  
1                         Bar  
2  Construction & Landscaping  
3                        Bank  
4           Electronics Store

Checking the Venues based on Neighborhood

In [94]:
venues_in_toronto.groupby('Neighborhood').head()

Neighborhood  \
0                                        Malvern, Rouge   
1                Rouge Hill, Port Union, Highland Creek   
2                Rouge Hill, Port Union, Highland Creek   
3                     Guildwood, Morningside, West Hill   
4                     Guildwood, Morningside, West Hill   
...                                                 ...   
1264  South Steeles, Silverstone, Humbergate, Jamest...   
1269  Clairville, Humberwood, Woodbine Downs, West H...   
1270  Clairville, Humberwood, Woodbine Downs, West H...   
1271  Clairville, Humberwood, Woodbine Downs, West H...   
1272  Clairville, Humberwood, Woodbine Downs, West H...   

      Neighborhood Latitude  Neighborhood Longitude  \
0                 43.806686              -79.194353   
1                 43.784535              -79.160497   
2                 43.784535              -79.160497   
3                 43.763573              -79.188711   
4                 43.763573              -79.188711   
...                     ...                     ...   
1264              43.739416              -79.588437   
1269              43.706748              -79.594054   
1270              43.706748              -79.594054   
1271              43.706748              -79.594054   
1272              43.706748              -79.594054   

                                                  Venue  \
0                                               Wendy’s   
1                                 Royal Canadian Legion   
2     SEBS Engineering Inc. (Sustainable Energy and ...   
3                                        RBC Royal Bank   
4                                     G & G Electronics   
...                                                 ...   
1264                                Sheriff's No Frills   
1269                                 Economy Rent A Car   
1270                                      Saand Rexdale   
1271                                          PC Garden   
1272                               Vectra Heavy Haulers   

                  Venue Category  
0           Fast Food Restaurant  
1                            Bar  
2     Construction & Landscaping  
3                           Bank  
4              Electronics Store  
...                          ...  
1264               Grocery Store  
1269         Rental Car Location  
1270                   Drugstore  
1271               Garden Center  
1272                  Truck Stop  

[430 rows x 5 columns]

So there are 430 records for each neighborhood.

Checking for the maximum venue categories

In [95]:
venues_in_toronto.groupby('Venue Category').max()

Neighborhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                Downsview East   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wine Shop                                        Church and Wellesley   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                              The Danforth West, Riverdale   

                    Neighborhood Latitude  Neighborhood Longitude  \
Venue Category                                                      
Accessories Store               43.718518              -79.464763   
Adult Boutique                  43.665860              -79.383160   
Airport                         43.737473              -79.394420   
Airport Food Court              43.628947              -79.394420   
Airport Gate                    43.628947              -79.394420   
...                                   ...                     ...   
Wine Bar                        43.653206              -79.379817   
Wine Shop                       43.665860              -79.383160   
Wings Joint                     43.628841              -79.520999   
Women's Store                   43.718518              -79.453512   
Yoga Studio                     43.679557              -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Wine Bar                          The National Club  
Wine Shop                                 Wine Rack  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[219 rows x 4 columns]

There are around 219 different types of Venue Categories. Interesting!

## One Hot encoding the venue Categories

In [96]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1268                  0               0        0                   0   
1269                  0               0        0                   0   
1270                  0               0        0                   0   
1271                  0               0        0                   0   
1272                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1268             0               0                0                 0   
1269             0               0                0                 0   
1270             0               0                0                 0   
1271             0               0                0                 0   
1272             0               0                0                 0   

      American Restaurant  Aquarium  ...  Truck Stop  \
0                       0         0  ...           0   
1                       0         0  ...           0   
2                       0         0  ...           0   
3                       0         0  ...           0   
4                       0         0  ...           0   
...                   ...       ...  ...         ...   
1268                    0         0  ...           0   
1269                    0         0  ...           0   
1270                    0         0  ...           0   
1271                    0         0  ...           0   
1272                    0         0  ...           1   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1268                              0                 0                      0   
1269                              0                 0                      0   
1270                              0                 0                      0   
1271                              0                 0                      0   
1272                              0                 0                      0   

      Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                   0         0          0            0              0   
1                   0         0          0            0              0   
2                   0         0          0            0              0   
3                   0         0          0            0              0   
4                   0         0          0            0              0   
...               ...       ...        ...          ...            ...   
1268                0         0          0            0              0   
1269                0         0       

Adding the neighbourhood to the encoded dataframe

In [97]:
toronto_venue_cat['Neighborhood'] = venues_in_toronto['Neighborhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0            0                  0               0        0   
1            0                  0               0        0   
2            0                  0               0        0   
3            0                  0               0        0   
4            0                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  Truck Stop  \
0                 0                    0  ...              0           0   
1                 0                    0  ...              0           0   
2                 0                    0  ...              0           0   
3                 0                    0  ...              0           0   
4                 0                    0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  
0                0         0          0            0              0  
1                0         0          0            0              0  
2                0         0          0            0              0  
3                0         0          0            0              0  
4                0         0          0            0              0  

[5 rows x 219 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood 

In [98]:
toronto_grouped = toronto_venue_cat.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Agincourt          0.0   
1                           Alderwood, Long Branch          0.0   
2  Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                  Bayview Village          0.0   
4                Bedford Park, Lawrence Manor East          0.0   

   Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                0.0             0.0      0.0                 0.0   
1                0.0             0.0      0.0                 0.0   
2                0.0             0.0      0.0                 0.0   
3                0.0             0.0      0.0                 0.0   
4                0.0             0.0      0.0                 0.0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0           0.0             0.0              0.0               0.0  ...   
1           0.0             0.0              0.0               0.0  ...   
2           0.0             0.0              0.0               0.0  ...   
3           0.0             0.0              0.0               0.0  ...   
4           0.0             0.0              0.0               0.0  ...   

   Train Station  Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0.0         0.0                            0.0               0.0   
1            0.0         0.0                            0.0               0.0   
2            0.0         0.0                            0.0               0.0   
3            0.0         0.0                            0.0               0.0   
4            0.0         0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                    0.0              0.0       0.0        0.0          0.0   
1                    0.0              0.0       0.0        0.0          0.0   
2                    0.0              0.0       0.0        0.0          0.0   
3                    0.0              0.0       0.0        0.0          0.0   
4                    0.0              0.0       0.0        0.0          0.0   

   Women's Store  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  

[5 rows x 219 columns]

Let's make a function to get the top most common venue categories

In [99]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [100]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [101]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village         Chinese Restaurant   
4                Bedford Park, Lawrence Manor East                Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Clothing Store        Breakfast Spot                Lounge   
1           Coffee Shop                   Gym          Skating Rink   
2                  Bank        Ice Cream Shop           Pizza Place   
3                  Bank                  Café   Japanese Restaurant   
4        Sandwich Place           Coffee Shop            Restaurant   

       5th Most Common Venue 6th Most Common Venue  \
0              Women's Store      Department Store   
1                        Pub        Sandwich Place   
2  Middle Eastern Restaurant        Sandwich Place   
3              Women's Store      Department Store   
4              Grocery Store    Italian Restaurant   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Eastern European Restaurant             Drugstore            Donut Shop   
1             Cuban Restaurant               Dog Run   Distribution Center   
2                  Bridal Shop                 Diner         Shopping Mall   
3  Eastern European Restaurant             Drugstore            Donut Shop   
4            Indian Restaurant      Sushi Restaurant                   Pub   

  10th Most Common Venue  
0                Dog Run  
1         Discount Store  
2                   Park  
3                Dog Run  
4             Hobby Shop

Let's make the model to cluster our Neighbourhoods

In [102]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [104]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=0, tol=0.0001, verbose=0)

Checking the labelling of our model

In [105]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0,
       0, 0, 1, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 4, 0, 4, 0, 0, 2, 4, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       4, 0, 0, 4, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
       0, 0, 0, 0, 3, 0, 0, 0, 4, 0, 0, 3], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories

In [106]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [107]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue  \
0 -79.194353             2.0  Fast Food Restaurant   
1 -79.160497             0.0                   Bar   
2 -79.188711             0.0            Restaurant   
3 -79.216917             0.0           Coffee Shop   
4 -79.239476             0.0       Thai Restaurant   

        2nd Most Common Venue 3rd Most Common Venue  \
0               Women's Store         Deli / Bodega   
1  Construction & Landscaping         Women's Store   
2         Rental Car Location            Donut Shop   
3       Korean BBQ Restaurant           Escape Room   
4                 Gas Station                Bakery   

         4th Most Common Venue        5th Most Common Venue  \
0  Eastern European Restaurant                    Drugstore   
1             Department Store            Electronics Store   
2            Electronics Store               Breakfast Spot   
3            Electronics Store  Eastern European Restaurant   
4                         Bank         Caribbean Restaurant   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                   Donut Shop               Dog Run   Distribution Center   
1  Eastern European Restaurant             Drugstore            Donut Shop   
2               Medical Center                  Bank    Mexican Restaurant   
3                    Drugstore            Donut Shop               Dog Run   
4           Athletics & Sports      Hakka Restaurant               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store                  Diner  
1               Dog Run    Distribution Center  
2          Intersection                    Spa  
3   Distribution Center         Discount Store  
4            Donut Shop       Department Store

Drop all the NaN values to prevent data skew

In [108]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [109]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [110]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [111]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1            Scarborough             0.0                   Bar   
2            Scarborough             0.0            Restaurant   
3            Scarborough             0.0           Coffee Shop   
4            Scarborough             0.0       Thai Restaurant   
6            Scarborough             0.0           Coffee Shop   
..                   ...             ...                   ...   
95             Etobicoke             0.0             Pet Store   
96            North York             0.0           Pizza Place   
99             Etobicoke             0.0           Pizza Place   
101            Etobicoke             0.0         Grocery Store   
102  Etobicoke Northwest             0.0   Rental Car Location   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1    Construction & Landscaping         Women's Store      Department Store   
2           Rental Car Location            Donut Shop     Electronics Store   
3         Korean BBQ Restaurant           Escape Room     Electronics Store   
4                   Gas Station                Bakery                  Bank   
6            Chinese Restaurant        Discount Store         Train Station   
..                          ...                   ...                   ...   
95                         Café            Beer Store          Liquor Store   
96       Furniture / Home Store      Cuban Restaurant             Drugstore   
99                  Coffee Shop            Playground    Chinese Restaurant   
101         Indie Movie Theater           Pizza Place        Sandwich Place   
102                   Drugstore         Garden Center            Truck Stop   

           5th Most Common Venue        6th Most Common Venue  \
1              Electronics Store  Eastern European Restaurant   
2                 Breakfast Spot               Medical Center   
3    Eastern European Restaurant                    Drugstore   
4           Caribbean Restaurant           Athletics & Sports   
6                    Bus Station                 Dance Studio   
..                           ...                          ...   
95             Convenience Store                  Coffee Shop   
96                    Donut Shop                      Dog Run   
99                Discount Store               Sandwich Place   
101                  Coffee Shop                   Beer Store   
102             Cuban Restaurant                   Donut Shop   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1                      Drugstore            Donut Shop               Dog Run   
2                           Bank    Mexican Restaurant          Intersection   
3                     Donut Shop               Dog Run   Distribution Center   
4               Hakka Restaurant               Dog Run            Donut Shop   
6    Eastern European Restaurant             Drugstore            Donut Shop   
..                           ...                   ...                   ...   
95                      Pharmacy                  Park           Pizza Place   
96           Distribution Center        Discount Store                 Diner   
99                  Intersection          Dance Studio            Donut Shop   
101         Fast Food Restaurant              Pharmacy           Gaming Cafe   
102                      Dog Run   Distribution Center        Discount Store   

    10th Most Common Venue  
1      Distribution Center  
2                      Spa  
3           Discount Store  
4         Department Store  
6                  Dog Run  
..                     ...  
95      College Auditorium  
96      Dim Sum Restaurant  
99                 Dog Run  
101                Dog Run  
102                  Diner  

[82 rows x 12 columns]

Cluster 2

In [112]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  North York             1.0        Baseball Field            Food Truck   
91   Etobicoke             1.0        Baseball Field         Women's Store   
97  North York             1.0        Baseball Field         Women's Store   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
32         Women's Store           Escape Room  Eastern European Restaurant   
91  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
97  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32             Drugstore            Donut Shop               Dog Run   
91             Drugstore            Donut Shop               Dog Run   
97             Drugstore            Donut Shop               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
32   Distribution Center         Discount Store  
91   Distribution Center         Discount Store  
97   Distribution Center         Discount Store

Cluster 3

In [113]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0   Scarborough             2.0  Fast Food Restaurant         Women's Store   
80         York             2.0  Fast Food Restaurant        Discount Store   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
0          Deli / Bodega  Eastern European Restaurant             Drugstore   
80        Sandwich Place                Women's Store     Electronics Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0             Donut Shop               Dog Run   Distribution Center   
80             Drugstore            Donut Shop               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
0         Discount Store                  Diner  
80   Distribution Center                  Diner

Cluster 4

In [114]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5              Scarborough             3.0            Playground   
23              North York             3.0                  Park   
40  East York/East Toronto             3.0                  Park   
98                    York             3.0         Jewelry Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5      Convenience Store         Women's Store          Dance Studio   
23     Convenience Store         Women's Store         Deli / Bodega   
40     Convenience Store           Coffee Shop         Women's Store   
98     Convenience Store         Women's Store         Deli / Bodega   

          5th Most Common Venue        6th Most Common Venue  \
5                     Drugstore                   Donut Shop   
23  Eastern European Restaurant                    Drugstore   
40                Deli / Bodega  Eastern European Restaurant   
98  Eastern European Restaurant                    Drugstore   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
5                Dog Run   Distribution Center        Discount Store   
23            Donut Shop               Dog Run   Distribution Center   
40             Drugstore            Donut Shop               Dog Run   
98            Donut Shop               Dog Run   Distribution Center   

   10th Most Common Venue  
5                   Diner  
23         Discount Store  
40    Distribution Center  
98         Discount Store

Cluster 5

In [115]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14        Scarborough             4.0      Asian Restaurant   
21         North York             4.0                  Park   
25         North York             4.0                  Park   
30         North York             4.0     Electronics Store   
44    Central Toronto             4.0    Photography Studio   
50   Downtown Toronto             4.0                  Park   
72         North York             4.0           Pizza Place   
74               York             4.0                  Park   
100         Etobicoke             4.0              Bus Line   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
14             Playground                  Park                 Intersection   
21          Women's Store          Dance Studio  Eastern European Restaurant   
25   Fast Food Restaurant     Food & Drink Shop                Women's Store   
30                Airport                  Park             Business Service   
44               Bus Line                  Park                  Swim School   
50             Playground                 Trail                Women's Store   
72                   Park            Smoke Shop                Women's Store   
74          Women's Store                  Pool                 Dance Studio   
100           Pizza Place                  Park               Sandwich Place   

    5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
14           Dance Studio                    Drugstore            Donut Shop   
21              Drugstore                   Donut Shop               Dog Run   
25          Deli / Bodega                    Drugstore            Donut Shop   
30          Deli / Bodega  Eastern European Restaurant             Drugstore   
44             Donut Shop                      Dog Run   Distribution Center   
50       Cuban Restaurant                   Donut Shop               Dog Run   
72       Cuban Restaurant                   Donut Shop               Dog Run   
74              Drugstore                   Donut Shop               Dog Run   
100         Women's Store                   Donut Shop               Dog Run   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14                Dog Run   Distribution Center         Discount Store  
21    Distribution Center        Discount Store                  Diner  
25                Dog Run   Distribution Center         Discount Store  
30             Donut Shop               Dog Run    Distribution Center  
44         Discount Store                 Diner     Dim Sum Restaurant  
50    Distribution Center        Discount Store                  Diner  
72    Distribution Center        Discount Store                  Diner  
74    Distribution Center        Discount Store                  Diner  
100   Distribution Center        Discount Store                  Diner

We have successfully cluster Toronto neighbourhood based on venue categories!